### PR TITLE
`cabal check`: add `--ignore` option

### LINKED ISSUE #8587: `cabal check`: add command line option to disable some warnings
With #8361 soon to be merged, `cabal check` will warn about missing upper bounds in all dependencies.
[As suggested in the discussion](https://github.com/haskell/cabal/issues/8291#issuecomment-1192913056), a way to disable this — or disable any specific warning, like `hlint` does — could be appreciated by some users.

### PR DESCRIPTION
**Pull request by ffaf1**
Add `--ignore` option to `cabal check`, so you can ignore specific warnings. Closes #8587.

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.
* [x] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added.

---

**QA Notes**

- cd into a new folder
- `cabal init -nm`
- `cabal check`
- inspect the output
- `cabal check --ignore=no-category`
- the specific warning is gone.

### CODE REVIEW

**Review by geekosaur — APPROVED** [review]

File: cabal-testsuite/PackageTests/Check/IgnoreWarning/cabal.test.hs
Comment (mpickering) [inline cabal-testsuite/PackageTests/Check/IgnoreWarning/cabal.test.hs:4]: Could we test some more scenarios here. 

* `--ignore` is passed an option which is not recognised, is it just ignored or fail?


Perhaps we should also enumerate all the possible errors here, so that if someone breaks the command line interface then we realise.

File: Cabal/src/Distribution/PackageDescription/Check/Warning.hs
Comment (mpickering) [inline Cabal/src/Distribution/PackageDescription/Check/Warning.hs:95]: I think the way this is displayed could be improved. 

Perhaps

```
ppExplanation ex ++ " [" ++ checkExplanationId ex ++ "]"
```

**Review by mpickering — DISMISSED** [review]
See my other comments

File: cabal-testsuite/PackageTests/Check/IgnoreWarning/cabal.test.hs
Comment (ffaf1) [reply cabal-testsuite/PackageTests/Check/IgnoreWarning/cabal.test.hs:4]: As now, an unrecognised flag is simply ignored. I agree an “unrecognised ignore” warning would be useful.

The problem I am facing now is that, for it to happen, I think I need to make `CheckExplanationID` a proper sum type, with as many constructors as `CheckExplanation`.

It seems some code duplication, but I do not see any way around that.

**Comment by ffaf1** [comment]
Ok, now:
- ignores are sandwitched between square brackets.
- ignores have their own names, and not names from constructors, e.g.

      Warning: [no-category] No 'category' field.

  (test on desirable ignores properties added).
- unrecognised ignores will emit a warning, not an error (test added).
- documentation has been updated with the list of ignores.

**Comment by mpickering** [comment]
I will look again tomorrow.

**Review by mpickering — COMMENTED** [review]
Looks good to me.

The only question I have is the value of having `CheckExplanationId` as a datatype rather than `type CheckExplanationId = String/ByteString/Text`. It seems that it is only converted to a string, and parsed from a string, are there any other manipulations of it that I am missing?

**Comment by ffaf1** [comment]
tl;dr: we need it to output a warning on unrecognised ignore strings.

To output a warning when an unrecognized ignore string is passed to cabal check — say a spelling error like `cabal check --ignore=msising-upper-bounds` — I need a set of all valid string values and then call ``"msising-upper-bounds" `elem` allCheckStrings``.

With `CheckExplanationId` as a datatype (instance of `Enum` and `Bounded`), I can build that set with `map ppCheckExplanationId ([minBound..maxBound] :: CheckExplanationId)`. Not only that, but when we add constructors to `CheckExplanation`, compiler warnings pops up if we forgot to add the relevant `CheckExplanationId`

If `CheckExplanationId` were to be a `String`, we could not use no more the `[minBound..maxBound]` trick; instead we would have to manually manage a `allCheckExplanationIds` list, and remember to modify it each time we modify or add a constructor to `CheckExplanation`. This is error prone as the compiler will not spot mismatches.

Does this make sense?

**Comment by ffaf1** [comment]
@mpickering does the explanation satisfy you? I am happy to answer questions or rework the code if needed.

**Comment by ffaf1** [comment]
@mpickering I will ping again. If there is anything I can do to help this review, I will happily do so.

**Comment by ulysses4ever** [comment]
@mpickering wasn't responsive recently. Maybe it's the holiday season... But given a very convincing explanation about the data type, I took the liberty to dismiss Matt's request for changes. We still need a second approval though...

**Comment by geekosaur** [comment]
FWIW @mpickering just said a few hours ago in #ghc:matrix.org that he's been very sick recently and not doing any work.

**Comment by ffaf1** [comment]
Thanks Artem and get well soon Matthew.

File: Cabal/src/Distribution/PackageDescription/Check/Warning.hs
Comment (ulysses4ever) [inline Cabal/src/Distribution/PackageDescription/Check/Warning.hs:302]: Maybe
```suggestion
-- can output a warning on unrecognised values.
-- NOTE: should be kept in sync with 'CheckExplanation'
```
or something like that.

To be honest, I'm not a fan of another data type with 137 constructors (and counting). But I don't have great ideas about it. Something fancier like `singletons` comes to mind...

1.  Keep `CheckExplanationID` as is.
2.  Change `CheckExplanation` to have one constructor with an extension point defined by a type family, e.g.

```
data CheckExplanation (id :: CheckExplanationID) =
  CE (Sing id) (CheckExplanationPayload id)

type family CheckExplanationPayload id where
  CheckExplanationPayload 'DuplicateSections = [UnqualComponentName]
  -- ... many more for every explanation with non-trivial payload
  CheckExplanationPayload id = () -- default payload is unit
```

Not sure it's the best way to skin this cat. Maybe someone with _working knowledge_ of Haskell beyond Haksell98 could suggest a better solution...

And if you change ChangeExplanation, you'll have to change a ton of code, I guess... So, I don't want to hold this PR hostage of this observation.

File: cabal-install/src/Distribution/Client/Setup.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/Setup.hs:1554]: Does it have to be `Generic`? Just your regular reminder that `Generic` is not free and can cost sensible compilation time.

File: cabal-install/src/Distribution/Client/Setup.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/Setup.hs:1573]: A little surprising that this patch changes the error message (to not mention the exit code anymore).

**Review by ulysses4ever — APPROVED** [review]
Great work, thank you! Nothing blocking from my perspective. So, it's up to you how to proceed.

File: cabal-install/src/Distribution/Client/Setup.hs
Comment (ulysses4ever) [reply cabal-install/src/Distribution/Client/Setup.hs:1573]: Sorry, not error message, but help message.

Comment (ffaf1) [reply cabal-install/src/Distribution/Client/Setup.hs:1573]: Well spotted, careless rebase.

File: cabal-install/src/Distribution/Client/Setup.hs
Comment (ffaf1) [reply cabal-install/src/Distribution/Client/Setup.hs:1554]: No need indeed, eliminated.

File: cabal-install/src/Distribution/Client/Check.hs
Comment (ffaf1) [reply cabal-install/src/Distribution/Client/Check.hs:62]: I will update the Haddock in various places and leave it synonymless.

File: Cabal/src/Distribution/PackageDescription/Check/Warning.hs
Comment (ffaf1) [reply Cabal/src/Distribution/PackageDescription/Check/Warning.hs:306]: This is an excellent suggestion. I have never actually used type familis, but they seem quite suited to the problem. Plan: merge this, then I will investigate how difficult it is to implement, report back on Matrix..

Comment (ffaf1) [reply Cabal/src/Distribution/PackageDescription/Check/Warning.hs:306]: Also: “keep in sync” invariant comes for free with `-WAll`.

Comment (ulysses4ever) [reply Cabal/src/Distribution/PackageDescription/Check/Warning.hs:306]: SGTM. Good to know about -Wall. Wouldn't hurt to spell it out in the docs anyway I think.

**Pull request merged**
